### DISCUSSION TITLE
ChemicalState.props() vs ChemicalProps(state): Different outputs

### DISCUSSION THREAD

**Discussion by volpatto** [reaktoro/reaktoro#382] [General]
Hi, Allan.

I'm not sure if the original idea of `state.props()` (`state` being an instance of `ChemicalState`) is to be different than `ChemicalProps(state)`.

The problem is: We have a problem that we have to rescale the species amount by a factor. I noticed that it was not been properly calculated (in our context) because we retrieve a bunch of properties using `scaled_state.props()` afterward. I noticed that the properties are the same as in the original `state`.

Then, I had a look in Reaktoro and found that the attribute `ChemicalState::props()` is initialized in the ChemicalState initialization and so it remains untouched by the scaling member functions (see [here](https://github.com/reaktoro/reaktoro/blob/5fa015d178f192e5410b6aa092cc7d533b003938/Reaktoro/Core/ChemicalState.cpp#L209)). While working with some post-processing calculations, I fell it a bit misleading, took me some time to figure it out what was happening and where was the problem. 

Is this suppose to behave this way? If negative, would it be convenient for you if I open a PR with a patch or you want to take this? The fix is quite straightforward, I could open a draft PR and you can polish it and merge, any way you like.

Here is an ipynb demonstrating the issue:

[ChemicalProps_QA.zip](https://github.com/reaktoro/reaktoro/files/15314942/ChemicalProps_QA.zip)

**Comment by allanleal** [comment]
Hi Diego, `ChemicalState::props()` works as a getter method to retrieve its internal `ChemicalProps` object, which is modified during equilibrium and kinetic calculations. It exists so that users can retrieve the final `ChemicalProps` object at the end of these calculations, rather than evaluating them again. 

Initially I evaluated the following alternative options:

* ChemicalProps object passed along with a ChemicalState object to the solve methods and updated at the end;
* ChemicalProps object retrieved from solver objects.

In the end, I decided to keep it in `ChemicalState` because it was more convenient in other parts of the library.

The reason why the internal object `ChemicalProps` is not updated with every change in `ChemicalState` (e.g. scaling operations, volume change of a phase, setting the quantity of a species, temperature/pressure change) is because these changes can be accumulated, without an expensive operation followed by each state alteration. Once user has completed all necessary state changes, the chemical properties can be updated.

You can then use:

```py
state.props().update(state)
```

or

```py
props = ChemicalProps(state)
```

or

```py
props = ChemicalProps(system)
props.update(state)
```

Let me know if this clarifies the reasons for the current behavior of `ChemicalState.props`.
